### ISSUE TITLE
visualize attention maps

### ISSUE THREAD

**Issue by zhaoxzz** [google-research/vision_transformer#18]
how to using rollout to visualize attention maps? thanks

**Comment by faustomorales** [comment]
Hi there! I'm not affiliated with the project, but I took a crack at this in a [Keras implementation](https://pypi.org/project/vit-keras/) I'm working on. That said, I'm not 100% sure I did this correctly. Feedback welcome! The visuals I'm getting out of it seem to make sense (see the [README](https://github.com/faustomorales/vit-keras/blob/master/README.md) for an example).

Here's a [direct link](https://github.com/faustomorales/vit-keras/blob/65724adcfd3979067ce24734f08df0afa745637d/vit_keras/visualize.py#L7-L45) to the implementing code.

**Comment by zhaoxzz** [comment]
@faustomorales thanks， it seems that in your code, attention weights didn't  add a identity matrix I to represent residual connection。and i find the original roll out code here [https://github.com/samiraabnar/attention_flow/blob/2272b01eb88d83eb368d841bb4e572330c592929/attention_graph_util.py#L104](url), but i didnot try it yet. thanks anyway

**Comment by faustomorales** [comment]
Thanks for that! I've amended the function to add the identity matrix to the weights.

**Comment by mjlm** [comment]
Hi @zhaoxzz, the code that we used to compute the attention rollout is below. As you said, the identity matrix needs to be added to account for the residual connections, although it doesn't make much of a difference visually if I remember correctly.

```python
def attn_rollout(attn_weights_list):
  eye = np.eye(attn_weights_list[0].shape[-1])
  attn_weights_rollout = [0.5 * attn_weights_list[0] + 0.5 * eye]
  for w in attn_weights_list[1:]:
    a = 0.5 * w + 0.5 * eye
    a_tilde = np.matmul(a, attn_weights_rollout[-1])
    attn_weights_rollout.append(a_tilde)
  return attn_weights_rollout
```

**Comment by zhaoxzz** [comment]
@mjlm thank you for your reply

**Comment by yurilla56** [comment]
@mjlm Hello, thank you for the explanation.
Could you pls give a hit how to get attn_weights?
Thank you!

**Comment by mjlm** [comment]
You can use `flax.nn.capture_module_outputs()` to capture the attention queries/keys/values like this:
```python
with flax.nn.capture_module_outputs() as outputs:
    predictions = model_def.call(params, imgs)
```

The code below generates a list of the names of the relevant variables captured in `outputs`:

```python
capture_names = []
for block in range(cfg.transformer.num_layers):
  for name in ['query', 'key', 'value']:
    capture_names.append(
        f'/Transformer/encoderblock_{block}/MultiHeadDotProductAttention_1/{name}')
```

From these variables, you can then compute the attention weights as in `flax.nn.dot_product_attention` (see [here](https://github.com/google/flax/blob/692a62c557c176e5ca22de9db7f935fce75ceb3c/flax/core/nn/attention.py#L39-L122)). Hope this helps!

**Comment by carlobretti** [comment]
@mjlm Hi! Thank you so much for your help! I tried getting the attention weights as you described, but if I just compute the attention weights as in `flax.nn.dot_product_attention`,  I always get the attention weights for each of the attention heads at each layer. I tried simply averaging all the heads for each layer before passing them to `attn_rollout`, but that doesn't seem to be working... Do you maybe have any other suggestions for doing this? Again, thank you so much for your amazing work!

**Comment by mjlm** [comment]
We also averaged the attention across all heads, so I'm not sure what's going wrong in your case. Below is the code used to compute the weights, and perform the rollout:

```python
from collections.abc import Iterable

import jax
from jax import lax
import jax.numpy as jnp
import numpy as np

NUM_LAYERS = 16


def get_dot_product_attention_weights(query,
                                      key,
                                      value,
                                      dtype=jnp.float32,
                                      bias=None,
                                      axis=None,
                                      precision=None):
  """Computes dot-product attention given query, key, and value.

  This function is adapted from https://github.com/google/flax/blob/master/flax/nn/attention.py#L37.

  Args:
    query: queries for calculating attention with shape of `[batch_size, dim1,
      dim2, ..., dimN, num_heads, mem_channels]`.
    key: keys for calculating attention with shape of `[batch_size, dim1, dim2,
      ..., dimN, num_heads, mem_channels]`.
    value: values to be used in attention with shape of `[batch_size, dim1,
      dim2,..., dimN, num_heads, value_channels]`.
    dtype: the dtype of the computation (default: float32)
    bias: bias for the attention weights. This can be used for incorporating
      autoregressive mask, padding mask, proximity bias.
    axis: axises over which the attention is applied.
    precision: numerical precision of the computation see `jax.lax.Precision`
      for details.

  Returns:
    Attention weights.
  """
  assert key.shape[:-1] == value.shape[:-1]
  assert (query.shape[0:1] == key.shape[0:1] and
          query.shape[-1] == key.shape[-1])

  if axis is None:
    axis = tuple(range(1, key.ndim - 2))
  if not isinstance(axis, Iterable):
    axis = (axis,)
  assert key.ndim == query.ndim
  assert key.ndim == value.ndim
  for ax in axis:
    if not (query.ndim >= 3 and 1 <= ax < query.ndim - 2):
      raise ValueError('Attention axis must be between the batch '
                       'axis and the last-two axes.')
  depth = query.shape[-1]
  n = key.ndim
  # batch_dims is  <bs, <non-attention dims>, num_heads>
  batch_dims = tuple(np.delete(range(n), axis + (n - 1,)))
  # q & k -> (bs, <non-attention dims>, num_heads, <attention dims>, channels)
  qk_perm = batch_dims + axis + (n - 1,)
  key = key.transpose(qk_perm)
  query = query.transpose(qk_perm)
  # v -> (bs, <non-attention dims>, num_heads, channels, <attention dims>)
  v_perm = batch_dims + (n - 1,) + axis
  value = value.transpose(v_perm)

  query = query / jnp.sqrt(depth).astype(dtype)
  batch_dims_t = tuple(range(len(batch_dims)))
  attn_weights = lax.dot_general(
      query,
      key, (((n - 1,), (n - 1,)), (batch_dims_t, batch_dims_t)),
      precision=precision)

  # apply attention bias: masking, droput, proximity bias, ect.
  if bias is not None:
    attn_weights = attn_weights + bias

  # normalize the attention weights
  norm_dims = tuple(range(attn_weights.ndim - len(axis), attn_weights.ndim))
  attn_weights = jax.nn.softmax(attn_weights, axis=norm_dims)
  attn_weights = attn_weights.astype(dtype)

  return attn_weights


# Compute attention weights:
get_dot_product_attention_weights_cpu = jax.jit(get_dot_product_attention_weights, backend='cpu')

attn_weights = []
for block in range(NUM_LAYERS):
  attn_weights.append(get_dot_product_attention_weights_cpu(
      query=outputs[f'/Transformer/encoderblock_{block}/MultiHeadDotProductAttention_1/query'],
      key=outputs[f'/Transformer/encoderblock_{block}/MultiHeadDotProductAttention_1/key'],
      value=outputs[f'/Transformer/encoderblock_{block}/MultiHeadDotProductAttention_1/value'],))

attn_weights = jax.tree_map(np.array, attn_weights)


# Get numpy attention weights:
attn_weights_np = [np.array(a) for a in attn_weights]
attn_weights_np_avg = [np.mean(a, axis=1, keepdims=True) for a in attn_weights_np]  # Average over heads


def attn_rollout(attn_weights_list):
  eye = np.eye(attn_weights_list[0].shape[-1])
  attn_weights_rollout = [0.5 * attn_weights_list[0] + 0.5 * eye]
  for w in attn_weights_list[1:]:
    a = 0.5 * w + 0.5 * eye
    a_tilde = np.matmul(a, attn_weights_rollout[-1])
    attn_weights_rollout.append(a_tilde)
  return attn_weights_rollout

attn_weights_np_avg_rollout = attn_rollout(attn_weights_np_avg)
```

**Comment by aoyanl** [comment]
> We also averaged the attention across all heads, so I'm not sure what's going wrong in your case. Below is the code used to compute the weights, and perform the rollout:
> 
> ```python
> from collections.abc import Iterable
> 
> import jax
> from jax import lax
> import jax.numpy as jnp
> import numpy as np
> 
> NUM_LAYERS = 16
> 
> 
> def get_dot_product_attention_weights(query,
>                                       key,
>                                       value,
>                                       dtype=jnp.float32,
>                                       bias=None,
>                                       axis=None,
>                                       precision=None):
>   """Computes dot-product attention given query, key, and value.
> 
>   This function is adapted from https://github.com/google/flax/blob/master/flax/nn/attention.py#L37.
> 
>   Args:
>     query: queries for calculating attention with shape of `[batch_size, dim1,
>       dim2, ..., dimN, num_heads, mem_channels]`.
>     key: keys for calculating attention with shape of `[batch_size, dim1, dim2,
>       ..., dimN, num_heads, mem_channels]`.
>     value: values to be used in attention with shape of `[batch_size, dim1,
>       dim2,..., dimN, num_heads, value_channels]`.
>     dtype: the dtype of the computation (default: float32)
>     bias: bias for the attention weights. This can be used for incorporating
>       autoregressive mask, padding mask, proximity bias.
>     axis: axises over which the attention is applied.
>     precision: numerical precision of the computation see `jax.lax.Precision`
>       for details.
> 
>   Returns:
>     Attention weights.
>   """
>   assert key.shape[:-1] == value.shape[:-1]
>   assert (query.shape[0:1] == key.shape[0:1] and
>           query.shape[-1] == key.shape[-1])
> 
>   if axis is None:
>     axis = tuple(range(1, key.ndim - 2))
>   if not isinstance(axis, Iterable):
>     axis = (axis,)
>   assert key.ndim == query.ndim
>   assert key.ndim == value.ndim
>   for ax in axis:
>     if not (query.ndim >= 3 and 1 <= ax < query.ndim - 2):
>       raise ValueError('Attention axis must be between the batch '
>                        'axis and the last-two axes.')
>   depth = query.shape[-1]
>   n = key.ndim
>   # batch_dims is  <bs, <non-attention dims>, num_heads>
>   batch_dims = tuple(np.delete(range(n), axis + (n - 1,)))
>   # q & k -> (bs, <non-attention dims>, num_heads, <attention dims>, channels)
>   qk_perm = batch_dims + axis + (n - 1,)
>   key = key.transpose(qk_perm)
>   query = query.transpose(qk_perm)
>   # v -> (bs, <non-attention dims>, num_heads, channels, <attention dims>)
>   v_perm = batch_dims + (n - 1,) + axis
>   value = value.transpose(v_perm)
> 
>   query = query / jnp.sqrt(depth).astype(dtype)
>   batch_dims_t = tuple(range(len(batch_dims)))
>   attn_weights = lax.dot_general(
>       query,
>       key, (((n - 1,), (n - 1,)), (batch_dims_t, batch_dims_t)),
>       precision=precision)
> 
>   # apply attention bias: masking, droput, proximity bias, ect.
>   if bias is not None:
>     attn_weights = attn_weights + bias
> 
>   # normalize the attention weights
>   norm_dims = tuple(range(attn_weights.ndim - len(axis), attn_weights.ndim))
>   attn_weights = jax.nn.softmax(attn_weights, axis=norm_dims)
>   attn_weights = attn_weights.astype(dtype)
> 
>   return attn_weights
> 
> 
> # Compute attention weights:
> get_dot_product_attention_weights_cpu = jax.jit(get_dot_product_attention_weights, backend='cpu')
> 
> attn_weights = []
> for block in range(NUM_LAYERS):
>   attn_weights.append(get_dot_product_attention_weights_cpu(
>       query=outputs[f'/Transformer/encoderblock_{block}/MultiHeadDotProductAttention_1/query'],
>       key=outputs[f'/Transformer/encoderblock_{block}/MultiHeadDotProductAttention_1/key'],
>       value=outputs[f'/Transformer/encoderblock_{block}/MultiHeadDotProductAttention_1/value'],))
> 
> attn_weights = jax.tree_map(np.array, attn_weights)
> 
> 
> # Get numpy attention weights:
> attn_weights_np = [np.array(a) for a in attn_weights]
> attn_weights_np_avg = [np.mean(a, axis=1, keepdims=True) for a in attn_weights_np]  # Average over heads
> 
> 
> def attn_rollout(attn_weights_list):
>   eye = np.eye(attn_weights_list[0].shape[-1])
>   attn_weights_rollout = [0.5 * attn_weights_list[0] + 0.5 * eye]
>   for w in attn_weights_list[1:]:
>     a = 0.5 * w + 0.5 * eye
>     a_tilde = np.matmul(a, attn_weights_rollout[-1])
>     attn_weights_rollout.append(a_tilde)
>   return attn_weights_rollout
> 
> attn_weights_np_avg_rollout = attn_rollout(attn_weights_np_avg)
> ```

hello， I want to ask a question, If I use this model for a multi-label classification task, how can I get the attention map of  each label?  Thank you very much!

**Comment by mjlm** [comment]
> hello， I want to ask a question, If I use this model for a multi-label classification task, how can I get the attention map of each label? Thank you very much!

In general, the attention maps are a function of the model parameters and the input image. The labels don't play a role in creating the maps, so there's no obvious way to get a map specific to each label, as far as I can tell. Maybe you can adapt a visualization technique like class activation mapping to your use case: http://cnnlocalization.csail.mit.edu/
